File: leetcode/easy/DesignHashSet.java
class MyHashSet {

    // --- Inner Node Class for Linked List ---
    // Represents a key in the hash set.
    // Made private static to encapsulate and avoid implicit reference to MyHashSet instance.
    private static class Node {
        final int key;
        Node next; // Pointer to the next node in case of collision

        public Node(int key, Node next) {
            this.key = key;
            this.next = next;
        }
    }

    // --- MyHashSet Fields ---
    private Node[] buckets; // Array of Node heads (each Node is the head of a linked list)
    private int capacity;          // Current capacity of the hash table (number of buckets)
    private int size;              // Current number of elements in the hash set
    private final double LOAD_FACTOR_THRESHOLD_UP = 0.75; // Threshold for resizing up
    private final double LOAD_FACTOR_THRESHOLD_DOWN = 0.25; // Threshold for resizing down (for shrink)
    private final int INITIAL_CAPACITY = 16; // Initial capacity, often a power of 2

    /**
     * Initializes a new instance of MyHashSet.
     */
    public MyHashSet() {
        this.capacity = INITIAL_CAPACITY;
        this.buckets = new Node[this.capacity]; // Initialize array of Node heads
        // Note: Each element in `buckets` is initially null. Nodes are added on `add`.
        this.size = 0;
    }

    /**
     * Computes the hash index for a given key.
     * Uses a simple modulo operation. Assumes non-negative keys for simplicity.
     * For negative keys, a more robust hash would be `(Integer.hashCode(key) & 0x7fffffff) % this.capacity`.
     *
     * @param key The key to hash.
     * @return The index in the buckets array.
     */
    private int hash(int key) {
        return key % this.capacity;
    }
    
    /**
     * Adds the specified element to this set if it is not already present.
     *
     * @param key The element to be added to this set.
     */
    public void add(int key) {
        int index = hash(key);
        Node current = this.buckets[index]; // Head of the linked list at this bucket

        // Traverse the linked list to check if the key already exists.
        while (current != null) {
            if (current.key == key) {
                return; // Key already exists, do nothing.
            }
            current = current.next;
        }

        // Key does not exist, add a new node at the beginning of the bucket's linked list.
        Node newNode = new Node(key, this.buckets[index]);
        this.buckets[index] = newNode;
        this.size++;

        // Check if resizing (expanding) is needed after adding a new entry.
        if ((double) this.size / this.capacity > LOAD_FACTOR_THRESHOLD_UP) {
            resize(this.capacity * 2);
        }
    }
    
    /**
     * Removes the specified element from this set if it is present.
     *
     * @param key The element to be removed from this set.
     */
    public void remove(int key) {
        int index = hash(key);
        Node head = this.buckets[index]; // Head of the linked list at this bucket

        // Use a dummy node to simplify removal, especially if the head of the list needs to be removed.
        Node dummy = new Node(-1, head); // Dummy node, key doesn't matter
        
        Node prev = dummy;
        Node current = head;

        // Traverse the linked list to find the node to remove.
        while (current != null && current.key != key) {
            prev = current;
            current = current.next;
        }

        // If current is not null, it means we found the node to remove.
        if (current != null) {
            prev.next = current.next; // Bypass the node to remove.
            this.buckets[index] = dummy.next; // Update the bucket head (in case head was removed).
            this.size--;

            // Check if resizing (shrinking) is needed after removing an entry.
            // Only shrink if capacity is not already at its minimum (INITIAL_CAPACITY).
            if (this.capacity > INITIAL_CAPACITY && (double) this.size / this.capacity < LOAD_FACTOR_THRESHOLD_DOWN) {
                resize(this.capacity / 2);
            }
        }
    }
    
    /**
     * Returns true if this set contains the specified element.
     *
     * @param key The element whose presence in this set is to be tested.
     * @return true if this set contains the specified element, false otherwise.
     */
    public boolean contains(int key) {
        int index = hash(key);
        Node current = this.buckets[index]; // Head of the linked list at this bucket

        // Traverse the linked list to find the key.
        while (current != null) {
            if (current.key == key) {
                return true; // Key found.
            }
            current = current.next;
        }
        return false; // Key not found.
    }

    /**
     * Resizes the hash table to a new capacity and rehashes all existing entries.
     *
     * @param newCapacity The new capacity for the hash table.
     */
    private void resize(int newCapacity) {
        // Store old buckets temporarily.
        Node[] oldBuckets = this.buckets;
        int oldCapacity = this.capacity;

        // Update capacity and initialize new buckets array.
        this.capacity = newCapacity;
        this.buckets = new Node[this.capacity];
        // No need to explicitly initialize each bucket to null, as it's the default for object arrays.
        
        this.size = 0; // Reset size, as add() will increment it again during rehash.

        // Rehash all entries from old buckets into new buckets.
        for (int i = 0; i < oldCapacity; i++) {
            Node node = oldBuckets[i];
            while (node != null) {
                this.add(node.key); // Use add method to rehash each entry
                node = node.next;
            }
        }
    }
}

File: leetcode/easy/DesignHashSet.py
from typing import Optional


class Node:
    """
    A node in a singly linked list. Used for chaining in the hash set.
    """
    def __init__(self, key: int, next_node: Optional['Node'] = None) -> None:
        self.key = key
        self.next = next_node


class MyHashSet:
    """
    A custom implementation of a HashSet using a hash table with separate chaining for collision resolution.
    The hash table dynamically resizes itself based on a load factor to maintain performance.
    """

    def __init__(self):
        """
        Initializes the HashSet.
        """
        # The ratio of size to capacity above which the hash table should grow.
        self.UP_LOAD_FACTOR = 0.75
        # The ratio of size to capacity below which the hash table should shrink.
        self.DOWN_LOAD_FACTOR = 0.25
        # The initial number of buckets in the hash table.
        self.capacity = 32
        # The number of unique keys currently stored in the hash set.
        self.size = 0
        # The list of buckets, where each bucket is the head of a linked list.
        self.buckets = [None] * self.capacity
        

    def add(self, key: int) -> None:
        """
        Adds a key to the hash set. If the key is already present, the set remains unchanged.
        This operation avoids a separate 'contains' check for better performance.
        """
        index = self.hash(key)
        curr = self.buckets[index]
        
        # Traverse the linked list in the bucket to check if the key already exists.
        while curr:
            if curr.key == key:
                return  # Key is already in the set, so do nothing.
            curr = curr.next
        
        # If the key is not found, add a new node to the head of the linked list.
        new_node = Node(key, self.buckets[index])
        self.buckets[index] = new_node
        self.size += 1
        
        # If adding the new key exceeds the upper load factor, resize to double the capacity.
        if (self.size / self.capacity) > self.UP_LOAD_FACTOR:
            self.resize(self.capacity * 2)

    def remove(self, key: int) -> None:
        """
        Removes a key from the hash set if it is present.
        This operation finds and removes the key in a single pass.
        """
        index = self.hash(key)
        
        # Use a dummy head to simplify the removal of the first node in the linked list.
        dummy = Node(0, self.buckets[index])
        curr = dummy
        key_found = False
        
        # Traverse the list to find the node to remove.
        while curr.next:
            if curr.next.key == key:
                # Remove the node by updating the 'next' pointer of the previous node.
                curr.next = curr.next.next
                key_found = True
                break
            curr = curr.next
            
        if key_found:
            # If a key was removed, update the bucket head and decrement the size.
            self.buckets[index] = dummy.next
            self.size -= 1
            # If the size drops below the lower load factor, resize to half the capacity.
            # We avoid shrinking below the initial capacity.
            if self.capacity > 32 and (self.size / self.capacity) < self.DOWN_LOAD_FACTOR:
                self.resize(self.capacity // 2)

    def contains(self, key: int) -> bool:
        """
        Returns True if the key is in the hash set, and False otherwise.
        """
        index = self.hash(key)
        curr = self.buckets[index]
        
        # Traverse the linked list at the calculated bucket index.
        while curr:
            if curr.key == key:
                return True
            curr = curr.next
        return False
    
    def resize(self, new_capacity: int):
        """
        Resizes the hash table to a new capacity and re-hashes all existing keys.
        """
        old_buckets = self.buckets
        
        # Reset the properties for the new table structure.
        self.size = 0
        self.capacity = new_capacity
        self.buckets = [None] * self.capacity

        # Re-add all keys from the old buckets to the new, resized buckets.
        for bucket_head in old_buckets:
            curr = bucket_head
            while curr:
                self.add(curr.key)
                curr = curr.next

    def hash(self, key: int) -> int:
        """
        Calculates the bucket index for a given key.
        """
        return key % self.capacity
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)
